### ISSUE TITLE
Evaluate PostgreSQL vs Elasticsearch for our use case; come up with better replication scheme if we stick with Elasticsearch

### ISSUE THREAD

**Issue by aldenstpage** [creativecommons/open-ledger#237]
- Is PostgreSQL full-text search sufficient for our use-case?
- If we keep using Elasticsearch, how can we improve the replication process?

**Comment by aldenstpage** [comment]
I set up PostgreSQL FTS with our full (10MM images) and got promising results. I was able to achieve good performance searching the entire database on my laptop. It's not as fast as Elasticsearch, but it is fast enough (10 to 50ms queries) and there is no separate indexing step, as opposed to Elasticsearch's 3 day indexing stage. See the psql-fulltext branch for details on how this is implemented.

Since our database will hold 300MM records in the near future, I want to make sure that PostgreSQL FTS still works well on this scale. I am working on a script that fills up our development database with an arbitrary amount of mock image data for the purposes of full text search.

Depending on the results of the load test, I will either migrate our old Elasticsearch queries to PostgreSQL or start implementing some of my ideas to improve our Elasticsearch replication pipeline.

**Comment by aldenstpage** [comment]
Postgres FTS works well on hundreds of millions of records. We could definitely make the switch if needed. However, I am concerned about how well this will scale to many concurrent users in the future. Sticking with Elasticsearch will give us more flexibility with scaling. It's also a lot more responsive when we're returning large numbers of search results at once, which the API may need to support. Issues with indexing speed can be addressed by using their bulk insert API properly and tuning configuration.

**TL;DR** Postgres FTS may not perform adequately when traffic spikes, as we have seen in the past when media outlets or social media pick up our projects. Let's stick with Elasticsearch.
